### ISSUE TITLE
Docker driver mount device support / documentation

### ISSUE THREAD

**Issue by KamilLelonek** [hashicorp/nomad#2671]
Nomad does not support (or does not document) how to [mount devices on a Docker container](https://docs.docker.com/engine/reference/commandline/run/#add-host-device-to-container---device).

A use case here being mounting the Nvidia GPUs to a docker container. The Docker syntax to do this is the `--device` switch. For example:

    docker run --device=/dev/nvidia0:/dev/nvidia0 cuda

Can this support please be added, or, if already present, documented?

**Comment by dadgar** [comment]
Do you have use for this outside of GPUs? First class GPU support will come.

**Comment by KamilLelonek** [comment]
The only use is to access Nvidia GPU.

**Comment by dadgar** [comment]
@KamilLelonek Okay I am going to close in favor of: https://github.com/hashicorp/nomad/issues/406.

This is being actively worked on!

**Comment by KamilLelonek** [comment]
What do you suggest for now? Should I use raw script with flags provided?

**Comment by dadgar** [comment]
@KamilLelonek Yep that is the best route to take for the time being!

**Comment by schmichael** [comment]
Closing as a duplicate of #406

**Comment by KamilLelonek** [comment]
Do you think that for such case at the moment `job.nomad` should contain the following:

    task "server" {
      driver = "exec"

      config {
        command = "docker run docker-registry/image-name:tag"
        args    = [
          "--device=/dev/nvidia0:/dev/nvidia0",
          "--device=/dev/nvidia1:/dev/nvidia1",
          "--device=/dev/nvidia2:/dev/nvidia2",
          "--device=/dev/nvidiactl:/dev/nvidiactl",
          "--device=/dev/nvidia-uvm:/dev/nvidia-uvm",
          "--device=/dev/nvidia-uvm-tools:/dev/nvidia-uvm-tools"
        ]
      }
    }

?

**Comment by dadgar** [comment]
I would do it with raw_exec but that should work. I would clean up the image afterwords!

**Comment by schmichael** [comment]
Just add `--rm` to cleanup the image on exit. And raw_exec will work much
better. You'll probably also need to run the task as a user with access to
docker. The easiest would be to just add the nobody user to the docker
group, but that gives docker access to any number of services.

**Comment by KamilLelonek** [comment]
Thanks, I need to change it anyway because I must provide only executable
and the rest of arguments as an array.

**Comment by schmichael** [comment]
Oops, I misspoke on `--rm` above. It should have been:

Just add `--rm` to cleanup the ~image~ container on exit. You'll need a script to also cleanup the image if desired.

**Comment by jaychris** [comment]
We need "--device" to mount usb devices and other serial control devices for hardware that container based applications are managing.

**Comment by pieterbreed** [comment]
I would like to have access to `/dev/fuse` inside my container that has an instance of `riofs` running. (An alternative to `s3fs`). To get this to work, I have to give both `privileged` and `--device /dev/fuse` on the docker command line. It would be nice if nomad supported devices in this way, over and above what #406 deals with.

**Comment by schmichael** [comment]
Allowing mounting arbitrary devices makes sense. I think the distinction
between GPUs (#406) and other devices is we intend for the scheduler to
have first class knowledge of GPU resources. Other devices will just
supported in driver configs.

Jay or Pieter: Would one of you mind opening a Github issue for your
"--device" use cases? It shouldn't be blocked by #406.
